Backport "Handle race in QueryStateTimer"

backport https://github.com/prestosql/presto/pull/841.  This should fix failures like


```
[ERROR] Tests run: 2240, Failures: 1, Errors: 0, Skipped: 2, Time elapsed: 2,808.075 s <<< FAILURE! - in TestSuite
[ERROR] tearDown(com.facebook.presto.execution.resourceGroups.db.TestQueuesDb)  Time elapsed: 5.906 s  <<< FAILURE!
java.lang.IllegalArgumentException: value is negative
	at io.airlift.units.Preconditions.checkArgument(Preconditions.java:26)
	at io.airlift.units.Duration.<init>(Duration.java:69)
	at io.airlift.units.Duration.succinctDuration(Duration.java:59)
	at io.airlift.units.Duration.succinctNanos(Duration.java:54)
	at com.facebook.presto.execution.QueryStateTimer.nanosSince(QueryStateTimer.java:237)
	at com.facebook.presto.execution.QueryStateTimer.nanosSince(QueryStateTimer.java:232)
	at com.facebook.presto.execution.QueryStateTimer.beginStarting(QueryStateTimer.java:93)
	at com.facebook.presto.execution.QueryStateTimer.beginRunning(QueryStateTimer.java:103)
	at com.facebook.presto.execution.QueryStateTimer.beginFinishing(QueryStateTimer.java:113)
	at com.facebook.presto.execution.QueryStateTimer.endQuery(QueryStateTimer.java:124)
	at com.facebook.presto.execution.QueryStateTimer.endQuery(QueryStateTimer.java:119)
	at com.facebook.presto.execution.QueryStateMachine.transitionToCanceled(QueryStateMachine.java:821)
	at com.facebook.presto.execution.SqlQueryExecution.cancelQuery(SqlQueryExecution.java:522)
	at java.util.Optional.ifPresent(Optional.java:159)
	at com.facebook.presto.execution.SqlQueryManager.cancelQuery(SqlQueryManager.java:463)
	at com.facebook.presto.tests.DistributedQueryRunner.cancelAllQueries(DistributedQueryRunner.java:463)
	at com.facebook.presto.tests.DistributedQueryRunner.close(DistributedQueryRunner.java:449)
	at com.facebook.presto.execution.resourceGroups.db.TestQueuesDb.tearDown(TestQueuesDb.java:86)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:217)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:712)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/presto-main/src/main/java/com/facebook/presto/execution/QueryStateTimer.java
+++ b/presto-main/src/main/java/com/facebook/presto/execution/QueryStateTimer.java
@@ -21,6 +21,7 @@ import java.util.Optional;
 import java.util.concurrent.atomic.AtomicReference;
 
 import static io.airlift.units.Duration.succinctNanos;
+import static java.lang.Math.max;
 import static java.util.Objects.requireNonNull;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 import static java.util.concurrent.TimeUnit.NANOSECONDS;
@@ -234,7 +235,7 @@ class QueryStateTimer
 
     private static Duration nanosSince(long start, long now)
     {
-        return succinctNanos(now - start);
+        return succinctNanos(max(0, now - start));
     }
 
     private Duration getDuration(AtomicReference<Duration> finalDuration, AtomicReference<Long> start)

